### PR TITLE
[DOPS-942] Makefile refactor

### PR DESCRIPTION
**Pull request by phoinixgrr**
#### Summary

This is a first proposal, on Makefile refactoring introducing the following improvements:

- Decoupling of Makefile Variables from actual targets
- Easy to develop locally, not requiring binaries on the actual dev machine (docker)
- Better makefile logging
- Dockerfile linting
- Auto generated Makefile help command
- Security improvements with the introduction of SBOM and Vulnerabilities reporting. Moreover `docker run` runs as a non privileged user
- Documentation with inline commenting
- Aims to be easily portable to other Mattermost Repositories

```bash
❯ make                                                                          
Usage:
make help                           to get help
make lint                           to lint all
make test                           to test all
make docker-build                   to build the docker image
make docker-push                    to push the docker image
make docker-sbom                    to print a sbom report
make docker-scan                    to print a vulnerability report
make docker-lint                    to lint the Dockerfile
make go-build                       to build binaries
make go-run                         to run locally for development
make go-test                        to run tests
make go-mod-check                   to check go mod files consistency
make go-update-dependencies         to update go dependencies (vendor)
make go-lint                        to lint go code
make go-doc                         to generate documentation
make clean                          to clean-up
```
Finally We also perform relevant changes to the `Dockerfile`

#### Ticket Link

Ticket: https://mattermost.atlassian.net/browse/DOPS-942

### CODE REVIEW

**Comment by pfltdv** [comment]
Amazing work @phoinixgrr.

**Comment by pfltdv** [comment]
Amazing work @phoinixgrr.

**Comment by pfltdv** [comment]
Shall we add `docker-sign` target to the makefile ? 

Cosign: https://github.com/sigstore/cosign (Keyless signing not publicly available at the moment https://github.com/sigstore/cosign/blob/main/KEYLESS.md)

What are your toughts about `docker-sign` ? @phoinixgrr @spirosoik

File: Makefile
Comment (spirosoik) [inline Makefile:12]: should we make the `--platform` configurable so we can create multi-platform docker builds? as you do for go binaries?

File: Makefile
Comment (spirosoik) [inline Makefile:14]: I think it will be better to have here by default the docker hub

**Comment by phoinixgrr** [comment]
> Shall we add `docker-sign` target to the makefile ?
> 
> Cosign: https://github.com/sigstore/cosign (Keyless signing not publicly available at the moment https://github.com/sigstore/cosign/blob/main/KEYLESS.md)
> 
> What are your toughts about `docker-sign` ? @phoinixgrr @spirosoik

Yes, Artifact signing is a requirement of DOPS-942. 

This aims to be a first iteration, on providing templated/improved basic functionality for our Projects. 
For sure, Cosign will be a part of the functionality provided in the next iteration

File: Makefile
Comment (spirosoik) [inline Makefile:26]: ideally we want along with @streamer45 to add `buildVersion` and `buildDate`

File: Makefile
Comment (spirosoik) [inline Makefile:35]: I really like this part. I mean I like it a lot 🚀

File: Makefile
Comment (spirosoik) [inline Makefile:80]: I think the part here we need to consider is that some people may haven't installed `awk` ever. if we have in mind that in parallel we want them to install less tooling (eg. move to docker) from now on we need to think how we can change this.

File: Makefile
Comment (spirosoik) [inline Makefile:97]: I like the formatting will help us a lot I believe in CI logging

File: Makefile
Comment (streamer45) [inline Makefile:173]: I didn't expect `go-lint` to run docker. Should we run the local binary instead? Either prompt to install it if missing or do it automatically.

File: Makefile
Comment (spirosoik) [inline Makefile:105]: is this `SPDX`? btw I think there is a flag `--format spdx-json` as JSON will be useful  and upload in artifacts of Github actions for example?

File: Makefile
Comment (pfltdv) [inline Makefile:19]: Using append(`+=`) mode for the variables is intended? For instance IMAGE_GOLINT make target will fail if user want to use his own image. Is it okey to use conditional assignment (`?=`) ?

```
ustafa@mmmmm:/srv/github/rtcd$ IMAGE_GOLINT=golangci/golangci-lint:v1.45.2 make go-lint
10:22:01 [ .. ] App linting...
docker: Error response from daemon: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "golangci/golangci-lint:v1.45.2@sha256:e84b639c061c8888be91939c78dae9b1525359954e405ab0d9868a46861bd21b": stat golangci/golangci-lint:v1.45.2@sha256:e84b639c061c8888be91939c78dae9b1525359954e405ab0d9868a46861bd21b: no such file or directory: unknown.
10:22:02 [FAIL]
make: *** [Makefile:173: go-lint] Error 1
```

File: Makefile
Comment (pfltdv) [inline Makefile:70]:
```suggestion
ERR	= echo ${TIME} ${RED}[FAIL]${CNone}
```

File: Makefile
Comment (streamer45) [inline Makefile:152]: How do you feel about printing out the command? Personally I am used to see which flags are passed.

File: Makefile
Comment (pfltdv) [inline Makefile:70]:
```suggestion
ERR	= echo ${TIME} ${RED}[FAIL]${CNone}
```

File: Makefile
Comment (pfltdv) [inline Makefile:71]:
```suggestion
OK	= echo ${TIME} ${GREEN}[ OK ]${CNone}
```

File: build/Dockerfile
Comment (spirosoik) [inline build/Dockerfile:15]: if we follow strict the linux directory structure we should this in `/opt` directory.
_`/opt` is a directory for installing unbundled packages (i.e. packages not part of the Operating System distribution, but provided by an independent source), each one in its own subdirectory_

For example have directory `/opt/rtcd` which includes the binary in `/opt/rtcd/bin/<binary>`

**Review by spirosoik — CHANGES_REQUESTED** [review]
Great job overall @phoinixgrr 🚀 

Few bullets for discussion:
- I think we miss the part of `docker manifest inspect/create/push` and then push the image
- I think we should still stick with `build`, `test`, `run`, `doc` target similar to lint. Each of them will use the same idea with lint.

**Review by pfltdv — COMMENTED** [review]
Editorial review.

File: Makefile
Comment (streamer45) [reply Makefile:12]: Realistically for the foreseeable future, rtcd will be Linux only (in production that is) as we rely on specific system level optimizations. If we want to support more platforms for development purposes I think we should still allow a way to override it so that a developer can just build for its own local environment.

**Review by streamer45 — COMMENTED** [review]
This is a great effort @phoinixgrr :raised_hands: I really like the _standardization_ process you started here. I only left a couple of questions/suggestions.

File: Makefile
Comment (phoinixgrr) [reply Makefile:19]: Nice. Haven't have thought this usage overwriting the configuration on runtime. 

Yes using conditional assignment would work in this case. But is this really needed? 
Is there a use case that the user of makefile would need to overwrite makefile variables, instead of directly editing the makefile and commiting the required changes? 

@streamer45  What is your take on this? Is there such a use case?

File: Makefile
Comment (phoinixgrr) [reply Makefile:26]: 👍 
Will also expose `buildVersion` & `buildDate`

File: Makefile
Comment (phoinixgrr) [reply Makefile:80]: Yes, I get the need for decoupling from tooling installed locally in users. 
This is the reason behind  designing `docker-lint` and `go-lint` to not depend on local tooling installed. 

Not sure if AWK is part of GNUtils, but I am pretty sure it is part of most -UNIX OSes. MacOS and Linux have it. 

Other than that, the only purpose for this target, is to dynamically format the `makefile help`, thus to produce this output: 
```
make help
Usage:
make help                           to get help
make lint                           to lint all
make test                           to test all
make docker-build                   to build the docker image
make docker-push                    to push the docker image
make docker-sbom                    to print a sbom report
make docker-scan                    to print a vulnerability report
make docker-lint                    to lint the Dockerfile
make go-build                       to build binaries
make go-run                         to run locally for development
make go-test                        to run tests
make go-mod-check                   to check go mod files consistency
make go-update-dependencies         to update go dependencies (vendor)
make go-lint                        to lint go code
make go-doc                         to generate documentation
make clean                          to clean-up
```

From my point of view:
- Lunching dockerd
- Waiting for linux image to download (containing awk + make) 
- Running the linux container mounting the local volume containing the makefile
- Running `make help` , to produce the output

Will introduce some delay in showing the user the actual makefile help. 

I guess we could format it under a docker container, Would you vote in favor of it?

File: Makefile
Comment (phoinixgrr) [reply Makefile:105]: Good point! Noting this down and will add it in the next iteration !

File: build/Dockerfile
Comment (phoinixgrr) [reply build/Dockerfile:15]: Agreed! Changing this to /opt!

File: Makefile
Comment (phoinixgrr) [reply Makefile:173]: Hey @streamer45 Thanks for commenting. Really appreciate your input on this! 

There is a requirement to decouple the targets, from local tooling installed in the developer workstation 
Do check this comment more context: https://github.com/mattermost/rtcd/pull/17#discussion_r859538715

File: Makefile
Comment (phoinixgrr) [reply Makefile:152]: From my point of view, I would like to see output only if a target fails to execute. 
Maybe we can check and see if we can add more verbosity options in the makefile, in a next iteration. Would that work for you?

File: Makefile
Comment (streamer45) [reply Makefile:19]: There are definitely use cases to override variables without going through the trouble of editing the file but it really depends on what the variable is about. Changing the image here may not be one of them.

Comment (phoinixgrr) [reply Makefile:19]: Ok, Let's resolve this for now, and let @streamer45  provide some feedback after using it for sometime. 
What do you say @pfltdv ?

**Comment by phoinixgrr** [comment]
Hey @agnivade. 
This is an initial effort on standardising the way we release.  
Ticket: https://mattermost.atlassian.net/browse/DOPS-908 

In this PR we start with tackling the Makefile. 

Your input is valuable, therefore we are adding you as a reviewer.

**Comment by phoinixgrr** [comment]
> Yeah, I had noticed this @phoinixgrr. Thanks for requesting a review!
> 
> For better visibility, once this is merged, I'd suggest to do a demo in R&D meeting. :)

Thanks @agnivade  
We actually going to perform a Design Review, and present it when we have a concrete proposal. 
This is just a part of the templating that we are proposing.

File: Makefile
Comment (pfltdv) [reply Makefile:19]: Yes lets wait for developers response, will resolve this thread.

File: Makefile
Comment (phoinixgrr) [reply Makefile]: Rebased and pushed changes

File: build/Dockerfile
Comment (phoinixgrr) [reply build/Dockerfile:15]: Rebased and pushed changes

File: Makefile
Comment (phoinixgrr) [reply Makefile]: Rebased and pushed changes

File: Makefile
Comment (phoinixgrr) [reply Makefile]: Rebased and pushed changes

File: Makefile
Comment (phoinixgrr) [reply Makefile]: Rebased and pushed changes

File: Makefile
Comment (streamer45) [inline Makefile:33]: I'd really like a way to override this so that a developer only has to compile for its local platform.

**Review by pfltdv — DISMISSED** [review]

File: Makefile
Comment (phoinixgrr) [reply Makefile]: Pushed https://github.com/mattermost/rtcd/pull/17/commits/1964aecde6da99cbd1c1ada56f6b011122f613a1 to assess this: 

Example for only building to `linux-amd64`
```
 GO_BUILD_PLATFORMS=linux-amd64 make go-build                       
11:35:07 [ .. ] go build rtcd-linux-amd64...
target="rtcd-linux-amd64"; \
	command="${target%%-*}"; \
	platform_ext="${target#*-}"; \
	platform="${platform_ext%.*}"; \
	export GOOS="${platform%%-*}"; \
	export GOARCH="${platform#*-}"; \
	echo export GOOS=${GOOS}; \
	echo export GOARCH=${GOARCH}; \
	CGO_ENABLED=0 \
	/opt/homebrew/bin/go build -mod=readonly -trimpath \
	-ldflags '-X "github.com/mattermost/rtcd/service.buildHash=393f0f7" -X "github.com/mattermost/rtcd/service.buildVersion=dev-393f0f7" -X "github.com/mattermost/rtcd/service.buildDate=2022-04-28 11:35"' \
	-o ./dist/rtcd-linux-amd64 \
	./cmd/rtcd || (echo `date +%H:%M:%S` [FAIL] && false)
export GOOS=linux
export GOARCH=amd64
11:35:08 [ OK ] go build rtcd-linux-amd64
```

Kudos to @pfltdv  🙇‍♂️

**Review by streamer45 — APPROVED** [review]
Thanks @phoinixgrr, looks great :

**Review by pfltdv — DISMISSED** [review]

**Review by pfltdv — DISMISSED** [review]

**Review by pfltdv — APPROVED** [review]

File: Makefile
Comment (spirosoik) [reply Makefile]: ok then no effort for multiplatform now. If we get requests we will revisit.

File: Makefile
Comment (spirosoik) [reply Makefile]: +1 🎸

**Review by spirosoik — APPROVED** [review]
great work @phoinixgrr

**Comment by agnivade** [comment]
@phoinixgrr - Sorry I didn't time to review this. In general, I don't have much opinions on the matter since I rarely use Makefile commands. I'll leave it to whatever you think best. :+1:

**Pull request merged**
